File: backend/domain/entities/order.py
"""
Order entity - Sales orders
"""
from datetime import datetime
from typing import Optional, List
from dataclasses import dataclass, field
from decimal import Decimal
from enum import Enum


class OrderType(str, Enum):
    """Order type enumeration"""
    SALE = "SALE"
    RETURN = "RETURN"


class PaymentMethod(str, Enum):
    """Payment method enumeration"""
    CASH = "CASH"
    BANK_TRANSFER = "BANK_TRANSFER"
    CREDIT = "CREDIT"
    MIXED = "MIXED"


class PaymentStatus(str, Enum):
    """Payment status enumeration"""
    PAID = "PAID"
    PARTIAL = "PARTIAL"
    UNPAID = "UNPAID"


@dataclass
class OrderItem:
    """
    Order item (line item) entity.
    
    Business Rules:
    - quantity must be > 0
    - unit_price must be >= 0
    - subtotal = quantity * unit_price - discount_amount
    """
    id: Optional[int]
    order_id: Optional[int]
    product_id: int
    unit_id: int
    quantity: Decimal
    unit_price: Decimal
    discount_percent: Decimal = Decimal("0")
    discount_amount: Decimal = Decimal("0")
    subtotal: Decimal = Decimal("0")
    line_total: Decimal = Decimal("0")
    notes: Optional[str] = None
    product_name: Optional[str] = None
    unit_name: Optional[str] = None
    created_at: Optional[datetime] = None
    
    def __post_init__(self):
        """Validate order item data"""
        if self.quantity <= 0:
            raise ValueError("Quantity must be greater than 0")
        
        if self.unit_price < 0:
            raise ValueError("Unit price cannot be negative")
        
        if self.discount_percent < 0 or self.discount_percent > 100:
            raise ValueError("Discount percent must be between 0 and 100")
    
    def calculate_subtotal(self) -> Decimal:
        """Calculate item subtotal"""
        base_amount = self.quantity * self.unit_price
        self.subtotal = base_amount - self.discount_amount
        self.line_total = self.subtotal
        return self.subtotal
    
    def apply_discount_percent(self):
        """Apply discount percentage to calculate discount amount"""
        if self.discount_percent > 0:
            base_amount = self.quantity * self.unit_price
            self.discount_amount = (base_amount * self.discount_percent) / 100


@dataclass
class Order:
    """
    Order entity representing sales orders.
    
    Business Rules:
    - total_amount = subtotal + tax_amount - discount_amount
    - debt_amount = total_amount - paid_amount
    - payment_status auto-updated based on paid_amount
    - Creates DEBT record if debt_amount > 0
    """
    id: Optional[int]
    owner_id: int
    order_code: str
    customer_id: int
    created_by: int
    order_date: datetime
    order_type: OrderType = OrderType.SALE
    subtotal: Decimal = Decimal("0")
    tax_rate: Decimal = Decimal("0")
    tax_amount: Decimal = Decimal("0")
    discount_amount: Decimal = Decimal("0")
    total_amount: Decimal = Decimal("0")
    paid_amount: Decimal = Decimal("0")
    debt_amount: Decimal = Decimal("0")
    payment_method: PaymentMethod = PaymentMethod.CASH
    payment_status: PaymentStatus = PaymentStatus.UNPAID
    notes: Optional[str] = None
    customer_name: Optional[str] = None
    customer_phone: Optional[str] = None
    draft_order_id: Optional[int] = None
    is_invoiced: bool = False
    is_accounted: bool = False
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    items: List[OrderItem] = field(default_factory=list)
    
    def __post_init__(self):
        """Validate order data"""
        if not self.order_code or len(self.order_code.strip()) == 0:
            raise ValueError("Order code is required")
        
        if self.tax_rate < 0 or self.tax_rate > 100:
            raise ValueError("Tax rate must be between 0 and 100")
        
        if self.paid_amount < 0:
            raise ValueError("Paid amount cannot be negative")
    
    def add_item(self, item: OrderItem):
        """Add item to order"""
        self.items.append(item)
        self.recalculate_totals()
    
    def recalculate_totals(self):
        """Recalculate order totals from items"""
        # Calculate subtotal from items
        self.subtotal = sum(item.calculate_subtotal() for item in self.items)
        
        # Calculate tax
        self.tax_amount = (self.subtotal * self.tax_rate) / 100
        
        # Calculate total
        self.total_amount = self.subtotal + self.tax_amount - self.discount_amount
        
        # Calculate debt
        self.debt_amount = self.get_debt_amount()
        
        # Update payment status
        self.update_payment_status()
    
    def update_payment_status(self):
        """Update payment status based on paid amount"""
        if self.paid_amount >= self.total_amount:
            self.payment_status = PaymentStatus.PAID
        elif self.paid_amount > 0:
            self.payment_status = PaymentStatus.PARTIAL
        else:
            self.payment_status = PaymentStatus.UNPAID
    
    def get_debt_amount(self) -> Decimal:
        """Calculate debt amount"""
        return max(Decimal("0"), self.total_amount - self.paid_amount)
    
    def has_debt(self) -> bool:
        """Check if order has outstanding debt"""
        return self.get_debt_amount() > 0
    
    def record_payment(self, amount: Decimal):
        """Record a payment for this order"""
        if amount <= 0:
            raise ValueError("Payment amount must be greater than 0")
        
        self.paid_amount += amount
        
        # Don't allow overpayment
        if self.paid_amount > self.total_amount:
            self.paid_amount = self.total_amount
        
        self.update_payment_status()
    
    def is_fully_paid(self) -> bool:
        """Check if order is fully paid"""
        return self.payment_status == PaymentStatus.PAID
    
    def can_be_invoiced(self) -> bool:
        """Check if order can be invoiced"""
        return len(self.items) > 0 and not self.is_invoiced

File: frontend/mobile/lib/models/order.dart
class Order {
  final int id;
  final String orderCode;
  final int? customerId;
  final double totalAmount;
  final double paidAmount;
  final String paymentStatus;
  final String createdAt;
  final List<OrderItem>? items;

  Order({
    required this.id,
    required this.orderCode,
    this.customerId,
    required this.totalAmount,
    required this.paidAmount,
    required this.paymentStatus,
    required this.createdAt,
    this.items,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: _parseInt(json['id']),
      orderCode: json['order_code'] ?? '',
      customerId: _parseInt(json['customer_id']),
      totalAmount: _parseDouble(json['total_amount']),
      paidAmount: _parseDouble(json['paid_amount']),
      paymentStatus: json['payment_status'] ?? 'PENDING',
      createdAt: json['created_at'] ?? '',
      items: json['items'] != null
          ? (json['items'] as List).map((i) => OrderItem.fromJson(i)).toList()
          : null,
    );
  }

  static double _parseDouble(dynamic value) {
    if (value == null) return 0.0;
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }

  static int _parseInt(dynamic value) {
    if (value == null) return 0;
    if (value is num) return value.toInt();
    if (value is String) return int.tryParse(value) ?? 0;
    return 0;
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'order_code': orderCode,
      'customer_id': customerId,
      'total_amount': totalAmount,
      'paid_amount': paidAmount,
      'payment_status': paymentStatus,
      'created_at': createdAt,
      'items': items?.map((i) => i.toJson()).toList(),
    };
  }
}

class OrderItem {
  final int id;
  final int productId;
  final String? productName;
  final double quantity;
  final double unitPrice;
  final double lineTotal;

  OrderItem({
    required this.id,
    required this.productId,
    this.productName,
    required this.quantity,
    required this.unitPrice,
    required this.lineTotal,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      id: _parseInt(json['id']),
      productId: _parseInt(json['product_id']),
      productName: json['product_name'],
      quantity: _parseDouble(json['quantity']),
      unitPrice: _parseDouble(json['unit_price']),
      lineTotal: _parseDouble(json['line_total']),
    );
  }

  static double _parseDouble(dynamic value) {
    if (value == null) return 0.0;
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }

  static int _parseInt(dynamic value) {
    if (value == null) return 0;
    if (value is num) return value.toInt();
    if (value is String) return int.tryParse(value) ?? 0;
    return 0;
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'product_id': productId,
      'product_name': productName,
      'quantity': quantity,
      'unit_price': unitPrice,
      'line_total': lineTotal,
    };
  }
}

class DraftOrder {
  final List<DraftItem> items;
  final String? rawText;

  DraftOrder({required this.items, this.rawText});

  factory DraftOrder.fromJson(Map<String, dynamic> json) {
    var rawItems = json['items'] as List? ?? [];
    return DraftOrder(
      items: rawItems.map((i) => DraftItem.fromJson(i)).toList(),
      rawText: json['raw_text'],
    );
  }
}

class DraftItem {
  final int productId;
  final String productName;
  final double quantity;
  final int unitId;
  final double unitPrice;

  DraftItem({
    required this.productId,
    required this.productName,
    required this.quantity,
    required this.unitId,
    required this.unitPrice,
  });

  factory DraftItem.fromJson(Map<String, dynamic> json) {
    return DraftItem(
      productId: _parseInt(json['product_id']),
      productName: json['product_name'] ?? 'Sản phẩm',
      quantity: _parseDouble(json['quantity']),
      unitId: _parseInt(json['unit_id']),
      unitPrice: _parseDouble(json['unit_price']),
    );
  }

  static double _parseDouble(dynamic value) {
    if (value == null) return 0.0;
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }

  static int _parseInt(dynamic value) {
    if (value == null) return 0;
    if (value is num) return value.toInt();
    if (value is String) return int.tryParse(value) ?? 0;
    return 0;
  }
}
